Source/WebCore/Modules/WebGPU/GPUDevice.cpp

```diff
@@ -59,27 +59,35 @@
 #include "GPUTextureDescriptor.h"
 #include "JSDOMPromiseDeferred.h"
 #include "JSGPUComputePipeline.h"
+#include "JSGPUDeviceLostInfo.h"
 #include "JSGPUInternalError.h"
 #include "JSGPUOutOfMemoryError.h"
 #include "JSGPUPipelineError.h"
 #include "JSGPURenderPipeline.h"
 #include "JSGPUValidationError.h"
+#include <pal/graphics/WebGPU/WebGPUAdapter.h>
 #include <wtf/IsoMallocInlines.h>
 
 namespace WebCore {
 
 WTF_MAKE_ISO_ALLOCATED_IMPL(GPUDevice);
 
-GPUDevice::GPUDevice(ScriptExecutionContext* scriptExecutionContext, Ref<PAL::WebGPU::Device>&& backing)
+GPUDevice::GPUDevice(ScriptExecutionContext* scriptExecutionContext, Ref<PAL::WebGPU::Device>&& backing, Ref<PAL::WebGPU::Adapter>& adapter)
     : ActiveDOMObject { scriptExecutionContext }
     , m_lostPromise(makeUniqueRef<LostPromise>())
     , m_backing(WTFMove(backing))
     , m_queue(GPUQueue::create(Ref { m_backing->queue() }))
     , m_autoPipelineLayout(createPipelineLayout({ { "autoLayout"_s, }, { } }))
+    , m_adapter(adapter)
 {
+    if (!adapter->isValid())
+        makeInvalid();
```

Comment:
I think the promise can be asynchronously resolved in `GPUDevice::lost(...)` to avoid having this logic here, I will give that a try.

Reply:
@litherum I moved the changes down into WebGPU.framework, leveraging what was already there. It seems to work, but needs more testing to make sure we don't retain the GPUDevice due to retaining it in `resolveDeviceLostPromise`.

Please let me know if you have any further comments, otherwise I will investigate the possible retain cycle prior to merging

Reply:
Oh I think I can just use `WeakPtr` instead of `Ref` in the callback, please let me know what you think.